apm-agent-core/src/main/java/co/elastic/apm/agent/bci/ElasticApmAgent.java

```diff
@@ -364,43 +364,50 @@ private static AgentBuilder applyAdvice(final ElasticApmTracer tracer, final Age
         final ElementMatcher<? super NamedElement> typeMatcherPreFilter = instrumentation.getTypeMatcherPreFilter();
         final ElementMatcher.Junction<ProtectionDomain> versionPostFilter = instrumentation.getProtectionDomainPostFilter();
         final ElementMatcher<? super MethodDescription> methodMatcher = new ElementMatcher.Junction.Conjunction<>(instrumentation.getMethodMatcher(), not(isAbstract()));
+        final AgentBuilder.RawMatcher matcher = new AgentBuilder.RawMatcher() {
+            @Override
+            public boolean matches(TypeDescription typeDescription, ClassLoader classLoader, JavaModule module, Class<?> classBeingRedefined, ProtectionDomain protectionDomain) {
+                if (classLoadingMatchingPreFilter && !classLoaderMatcher.matches(classLoader)) {
+                    return false;
+                }
+                if (typeMatchingWithNamePreFilter && !typeMatcherPreFilter.matches(typeDescription)) {
+                    return false;
+                }
+                boolean typeMatches;
+                try {
+                    typeMatches = typeMatcher.matches(typeDescription) && versionPostFilter.matches(protectionDomain);
+                } catch (Exception ignored) {
+                    // could be because of a missing type
+                    typeMatches = false;
+                }
+                if (typeMatches) {
+                    logger.debug("Type match for instrumentation {}: {} matches {}",
+                        instrumentation.getClass().getSimpleName(), typeMatcher, typeDescription);
+                    try {
+                        instrumentation.onTypeMatch(typeDescription, classLoader, protectionDomain, classBeingRedefined);
+                    } catch (Exception e) {
+                        logger.error(e.getMessage(), e);
+                    }
+                    if (logger.isTraceEnabled()) {
+                        logClassLoaderHierarchy(classLoader, logger, instrumentation);
+                    }
+                }
+                return typeMatches;
+
+            }
+        };
         return agentBuilder
-            .type(new AgentBuilder.RawMatcher() {
+            .type(logger.isDebugEnabled() ? new AgentBuilder.RawMatcher() {
```

Comment:
Yep, that's a good idea 👍

Reply:
Another style-related idea - if we keep `matcher` as a field, maybe also use `statsCollectionMatcher` as well instead of creating this anonymous type.

Reply:
Sorry, I do not really understand your comment. Do you want me to make the variable matcher a field of the new AgentBuilder.RawMatcher?

Reply:
No, sorry for not being very clear.
I think this statement above would be more readable if it looks something like that:
```
.type(instrumentationStats.shouldTimeMatching() ? statsCollectionMatcher : matcher);
```
and the way to achieve that is create a `statsCollectionMatcher` field, like you did for `matcher`.

Reply:
Done. Thank You for the clarification.

Reply:
Thanks! Exactly what I meant, even though I wrote "field", where it's actually a variable, which is probably what mislead you 😊